### ISSUE TITLE
Internal "unsafe fn" is added to global namespace

### ISSUE THREAD

**Issue by RalfJung** [ogoffart/unsafe_fn#1]
If I read the macro code correctly, the internal `__unsafe_fn_<foo>` is added to the global namespace, on the "outside" of the function. Why is that? Wouldn't it be better to hide it inside, i.e., to desugar to something like
```rust
unsafe fn foo(...) {
  __unsafe_fn_foo(...);
  fn __unsafe_fn_(...) { /* body */
}
```

**Comment by ogoffart** [comment]
I had done that originally, but it did not work for function that have self.

The problem is what type to give to self in the inner function:

```
unsafe fn foo(self) {
  __unsafe_fn_foo(self);
  fn __unsafe_fn_foo(self : /* What Type Here? */) { /* body with self replaced by self_ */
}
```
 `Self` does not work as a type because of the `error[E0401]: can't use generic parameters from outer function`
So that's why i need to forward to a sibling version.

The version 0.1.1 of the crate now only forward to a sibling version only if the usage of `self` or `Self` is detected, and use a inner function otherwise.

If you know a way to have it work for self with a inner function, let me know

**Comment by RalfJung** [comment]
Ah good point, it would have to also copy the generic parameters of the `impl` block. I do not think that is possible.

But I think this means `#[unsafe] fn` won't work very well inside `trait` or `impl Trait for`.

**Comment by ogoffart** [comment]
> it would have to also copy the generic parameters of the impl block. I do not think that is possible.

Right, that's why if one use `self` or `Self` the body is made as a sibling functions instead. Which means they are added to the trait or the struct,  which is not ideal, but it is still a private undocumented function.  (I was thinking about adding some random identifier or hash in the function name, so one can be sure nobody can call it.)

> But I think this means #[unsafe] fn won't work very well inside trait or impl Trait for.

Could you provide examples of this?

Also, if `#[unsafe]` were a compiler attribute, this would not be a problem.

**Comment by RalfJung** [comment]
I mean something like
```rust
trait Foo {
  #[unsafe] fn foo(&self) {
    // unsafe trait fn with a default impl with a safe body
  }
}
```
will actually add a `__unsafe_fn_foo` to the trait itself. But I guess it has a default impl so that's not a big deal...

However, with
```
trait Foo { unsafe fn foo(&self); }

impl Foo for u32 {
  #[unsafe] fn foo(&self) { ... }
}
```
I think there will be errors saying that `__unsafe_fn_foo` is not a trait item.

**Comment by ogoffart** [comment]
Ah right, the trait implementation in `impl Foo for Yyy` will not work.

Unless  `trait Foo { #[unsafe_fn] fn foo(&self); }`  also declares  a `__unsafe_fn_foo` with a defaulted body. But that's less than ideal.

So one way I see it is that I'm gonna have to transform all the possible paths in the body and put them as generic argument of the inner function.
